olive/engine/footprint.py

```diff
@@ -0,0 +1,189 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import logging
+from collections import OrderedDict
+from datetime import datetime
+from pathlib import Path
+
+from olive.common.config_utils import ConfigBase, config_json_dumps, config_json_loads
+
+logger = logging.getLogger(__name__)
+
+
+class FootprintNodeMetric(ConfigBase):
+    """
+    value: {"metric_name": metrics_value, ...}
+    cmp_direction: will be auto suggested. The format will be like: {"metric_name": 1, ...},
+        1: higher is better, -1: lower is better
+    is_goals_met: if the goals set by users is met
+    """
+
+    value: dict
+    cmp_direction: dict = None
+    is_goals_met: bool = False
+
+
+class FootprintNode(ConfigBase):
+    # None for no parent which means current model is the input model
+    parent_model_id: str = None
+    model_id: str
+    model_config: dict = None
+    from_pass: str = None
+    pass_run_config: dict = None
+    is_pareto_frontier: bool = False
+    # TODO add EP/accelerators for same_model_id metrics
+    metrics: FootprintNodeMetric = FootprintNodeMetric(value={}, cmp_direction={}, is_goals_met=False)
```

Comment:
I think this has the same potential issue as using {}/[] as default value. Can `metric` also be `None` by default? Looking at the other parts of the code (`resolve_metric`, `get_candidates`), it looks like you already assume the default value of `metric` is `None`.

Reply:
I think it is fine for parameter to accept {}, not the default value. But anyway, let me remove this. Thanks. :)

Reply:
I just checked how pydantic models behave with defaults and like you said, it doesn't appear to have the risk of using mutable object as default. Thanks for updating regardless!